doc/crdt-details.md

```diff
@@ -0,0 +1,1054 @@
+    delta.apply(&self.text)
+}
+```
+
+#### Engine::deletes_from_cur_union_for_index
+
+If you look back at the [example history scenario](#example-history), you'll see `back_computed_deletions_from_6_union`, which shows that for any past `Revision` we can find a set of deletions from the current union string that result in the past text. This helper is what computes deletion sets like `back_computed_deletions_from_6_union`.
+
+We can find an `old_deletes_from_cur_union` by taking our current `deletes_from_union` and walking backwards through our list of `Revision`s undoing the changes they have made since the old revision.
+
+There's a problem with this though, the `inserts` and `deletes` subsets use indices in the coordinate space of the union string at the time the `Revision` was created, which may be smaller than our current union string.
+
+We could keep track of the transform and account for it, but it's easier to use a helper we need anyway elsewhere that computes what the `deletes_from_union` actually would have been at that previous point in time, that is, relative to the old union string not the current one like we want. This helper is called `Engine::deletes_from_union_for_index` and it performs the work of un-deleting and un-undoing everything after the old revision.
+
+Then we can take this `old_deletes_from_union` and `Subset::transform_union` it through the `inserts` since that old revision to get the `old_deletes_from_cur_union` we wanted. This puts it in the right coordinate space and the `union` part of `transform_union` makes sure that new inserts are considered as deleted (so not present) in `old_deletes_from_cur_union`.
+
+```rust
+/// Get the Subset to delete from the current union string in order to obtain a revision's content
+fn deletes_from_cur_union_for_index(&self, rev_index: usize) -> Cow<Subset> {
+    let mut deletes_from_union = self.deletes_from_union_for_index(rev_index);
+    for rev in &self.revs[rev_index + 1..] {
+        if let Edit { ref inserts, .. } = rev.edit {
+            if !inserts.is_empty() {
+                deletes_from_union = Cow::Owned(deletes_from_union.transform_union(inserts));
+            }
+        }
+    }
+    deletes_from_union
+}
+```
+
+#### Engine::deletes_from_union_for_index
+
+This function uses the property that each `Revision` contains the information necessary to reverse it to work backwards from the current state of `deletes_from_union` to the past state. For every `Edit` revision it `subtract`s the `deletes` (meaning if something was deleted twice, this will only reverse one), but only if they weren't undone, and then uses `transform_shrink` to reverse the coordinate transform of the `inserts` so that the indices in the intermediate `old_deletes_from_union` refer to the previous union string. `Undo` edits store the symmetric differences of the `deletes_from_union` and the currently undone groups, so those are just reversed.
```

Comment:
> This function uses the property that each `Revision` contains the information necessary to reverse it to work backwards from the current state of `deletes_from_union` to the past state.

"reverse it to": maybe needs a comma after "it"? or make "it" "itself"? or "in order to"?